File: usaco/python/2020_open_soical_distancing/socdist.py
import sys

with open("socdist.in", "r") as fin:
  lines = fin.readlines()

n, m = map(int, lines[0].split())
intervals = []
for i in range(m):
    intervals.append(list(map(int, lines[1+i].split())))

intervals.sort()

def verify(gap, n, m, intervals):
  count = 0
  prev = 0
  for i in range(m):
    if i ==  0 or (intervals[i][0] - prev) >= gap:
      increased = (intervals[i][1] - intervals[i][0])//gap
      prev = intervals[i][0] + increased * gap
      count += increased + 1
    else:
      increased = (intervals[i][1] - prev)//gap
      prev += increased * gap
      count += increased
  if count >= n:
    return True
  else:
    return False

left = 0
right = intervals[m-1][1] - intervals[0][0]

result = 0
while left <= right:
  middle = left + (right - left)//2
  if verify(middle, n, m, intervals):
    left = middle + 1
    result = middle
  else:
    right = middle - 1

with open('socdist.out', 'w') as fout:
  fout.write(str(result) + '\n')

File: usaco/cpp/2020_open_social_distancing/socdist.cpp
#include <iostream>
#include <fstream>
#include <vector>
//#include <utility>
#include <algorithm>

using namespace std;

int n, m;
vector<pair<long long, long long>> intervals;
//    vector<int[2]> intervals;

bool verify(long long gap)
{
    int count = 0;
    long long prev = 0;
    for (int i = 0; i < intervals.size(); i++)
    {
        if (i == 0 || (intervals[i].first - prev) >= gap)
        {
            int increased = (intervals[i].second - intervals[i].first)/gap;
            prev = intervals[i].first + increased * gap;
            count += increased + 1;
        }
        else
        {
            int increased = (intervals[i].second - prev)/gap;
            prev += increased * gap;
            count += increased;
        }
        if (count >= n)
        {
            return true;
        }
    }
    return false;
}

int main()
{
    ifstream fin("socdist.in");
    ofstream fout("socdist.out");

    fin >> n >> m;
    cout << "n=" << n << endl;
    cout << "m=" << m << endl;

    intervals.resize(m);

    for (int i = 0; i < m; i++)
    {
        fin >> intervals[i].first >> intervals[i].second;
        //cout << "intervals[" << i << "].first=" << intervals[i].first << ", intervals[" << i << "].second=" << intervals[i].second << endl;
        //fin >> intervals[i][0] >> intervals[i][1];
        //cout << "intervals[" << i << "][0]=" << intervals[i][0] << ", intervals[" << i << "][1]=" << intervals[i][1] << endl;
    }

    sort(intervals.begin(), intervals.end());

    int left = 0;
    int right = intervals[m-1].second - intervals[0].first;
    int result = left;

    while (left <= right)
    {
        int middle = left + (right - left)/2;
        if (verify(middle))
        {
            left = middle + 1;
            result = middle;
        }
        else
        {
            right = middle - 1;
        }
    }

    fout << result << endl;

    return 0;
}